### ISSUE TITLE
"You need to login first. There is no valid certificate"

### ISSUE THREAD

**Issue by monhustla** [carpedm20/LINE#115]
I have made a nice bot with this that works well. How long are y'all able to keep your programs up without rerunning it? I have to restart the program every hour or I get "You need to login first. There is no valid certificate."

**Comment by R-N** [comment]
24/7. It runs "forever" now.

That error message is from updateAuthToken failing to do its job, because theres no certificate. It seems like you logged in with an authToken.

How did you get your authToken? When you get authToken, it also creates a ".line.crt" file. Place that inside your .py directory and inside your python binary directory (not sure which will be used. just do both lol)

If you're hosting it in heroku, it cant create that file. You'll have to edit api.py to save the certificate in a heroku global variable or in a table in postgresql, instead of writing it in ".line.crt", and load certificate from where you saved it. The certificate is a string/text.

When you login with authToken only, your client is missing :
-id (email)
-password
-certificate
-provider

You may want to provide them. Just add them like
client.id = " asd@gmail.com"
see api.py for setting client.provider

**Comment by R-N** [comment]
The "provider" field tells the client if your id is an id or an email

**Comment by monhustla** [comment]
Hey, thanks for the update. A lot of good info there. I'm running the beta
of my bot in cmd. I have it login with the email and pw associated with the
account. Client variable stays equal to that when the code reruns every 3
minutes. The program is saved in my LINE folder within python27, the
certificate is there as well.

Can I pm you a picture of how I have it setup (it's a very small one page
program) to see if you can see a quick fix? I appreciate the help.

On Sat, Jun 3, 2017 at 3:23 AM, Muhammad Rizqi Nur <notifications@github.com

**Comment by monhustla** [comment]
Here I'll just explain it here. So it runs in cmd, the program is saved in
the line directory within python 27, same with my certification. Code
framework is:

client=(email,pw)

s=sched.scheduler(time.time, time.sleep)
def do_something:
        if xxxx=true:
                friend=client.contacts
                friendsendMessage(....)
        s.enter(timed schedule, 1, do_something, (sc)
s.enter(timed schedul, 1, do_something(s,))
s.run()


That's what I have it set to now, I've also tried having it login again
within the if statement (only email and pw).  With both, it will work for a
few true statements and then eventually give me the error. Do I need to
remove or move the login from somewhere or possibly save auth within the
program?

**Comment by R-N** [comment]
Oh, right. 
I remember that api.py sets client's certificate to whatever it gets when it gets a certificate message. There are two of them if I remember correctly. Sometime it receives an empty certificate and overwrite the old one (or maybe it's just me, cuz my connection sucks)

Make it check if its an empty string (or None) or not before setting.
 
Also add in LineClient's __init__ to set self.certificate to an empty string or whatever. Just to make sure the variable exists, so it doesnt throw more errors.

Anyway, the client should always have certificate if you are able to login (email/id&pass login creates certs, authToken login needs certs). So the issue must be "the cert got deleted somehow after it logged in"

Sorry my previous comment was wrong. AuthToken login is not missing the certificate, it needs certificate

**Comment by monhustla** [comment]
This program is saved in the same directory as my certificate. I even added
my certificate to my path.

The basis of my program is this: it checks a forum for a new message from
someone. If a new message is found, it sends the message to every friend
and every group I'm in. The bot sends about 10-15 messages in one push. If
the forums are very active it will send maybe 100 messages an hour.

Is it that I'm pushing too many messages? And with the code framework I
detailed above, do you see any reason why that wouldn't work?

On Sun, Jun 4, 2017 at 9:11 AM, Muhammad Rizqi Nur <notifications@github.com

**Comment by monhustla** [comment]
The certificate is here, saved. It hasn't been altered in 3 days.

[image: Inline image 1]

**Comment by monhustla** [comment]
And you're talking right here in api.py right? (sorry about all of the
messages)

[image: Inline image 1]

**Comment by monhustla** [comment]
The error is always from models.py line 68, in sendMessage raise e

Exception: Error: You need to login first. There is no valid certificate.

**Comment by monhustla** [comment]
[image: Inline image 1]

**Comment by monhustla** [comment]
Okay this is very weird. So I tested today running code on an 80 second interval sending a "test" message to only friends. It worked fine every 80 seconds for a full hour. I then went back to my program for sending blast messages to groups. It sent the message to all of the groups and then signed me out. What is going on here?

**Comment by R-N** [comment]
Yes, the .line.crt isnt altered.

What I meant is the certificate in the LineClient object. self.certificate
There are two piece of code in apy.py which sets that, one of them writes .line.crt, one doesnt. 

I dont think it throws certificate error in sendMessage. Must be a chain from updateAuthToken. (Try-except-raise). I remember it will try to updateAuthToken if it gets a TalkException that means the authToken has expired

**Comment by R-N** [comment]
Yep. sendMessage in models.py has these lines :

except Exception as e:
        raise e

God knows what it's for.

**Comment by monhustla** [comment]
Here is the framework for the code:



This same code allows messages to contacts to be looped. However, when I
change it to groups, like above, it sends the message and then doesn't loop
giving the error from line 68 of models.py.

Why does sendmessage allow contacts to be looped but not groups? I'm trying
to have my bot send messages to the groups it is in.

**Comment by R-N** [comment]
Try to search for the error message in this api, only updateAuthToken throws that

See what caused it, and what may caused that cause.

In my case it was that the LineClient.certificate became empty somehow after I logged in. No, .line.crt isnt altered at all.

**Comment by monhustla** [comment]
Here's the code that code: 
![img_0714](https://cloud.githubusercontent.com/assets/20822555/26765600/a958a42e-4944-11e7-8456-802ceba26ee3.JPG)

**Comment by R-N** [comment]
Have you tried my suggestion though?
To check if the cert message isnt empty first before setting that as its certificate in api.py

**Comment by monhustla** [comment]
@R-N That's the thing, I don't have to restart pin validation everytime I run the program. I can go to cmd and simply type in "file".py and it reruns.

**Comment by monhustla** [comment]
Here is what my API.py looks like 
![img_0715](https://cloud.githubusercontent.com/assets/20822555/26765619/fef31dce-4944-11e7-9b98-eb3486b7e2b7.JPG)

**Comment by R-N** [comment]
Let me make myself clear,

I dont think it has anything to do with how you use the api. It's a problem within the api itself.

The fault isnt in your code.

**Comment by monhustla** [comment]
I  know bud, I just don't know exactly how to update api.py to make it work to send messages to groups. I think I blasted too many group messages the first day I got it running and that's why it's stopped. I didn't even know what python was until last week.

**Comment by monhustla** [comment]
![img_0716](https://cloud.githubusercontent.com/assets/20822555/26765632/60efc504-4945-11e7-8533-a02c4a18dfff.JPG)

**Comment by R-N** [comment]
Wait, so you're logging in with authToken afterall. Why would you even do that (recreating the client)

Try putting this after the "client=LineClient(authToken=authToken)"

print(client.certificate)

Also have you tried my damn suggestion yet? The checking cert thing.

Look for "self.certificate =" in api.py
Make sure it doesnt set self.certificate to empty string or None

**Comment by monhustla** [comment]
So @R-N , do I just "certificate= .line.crt" in the api.py? or do I keep it empty as ""?

**Comment by monhustla** [comment]
Sorry buddy, I'm awful with this stuff :/

**Comment by R-N** [comment]
Oh sorry, you couldve said that youbdidnt understand earlier lol.

You blasting messages make the authToken expire. The client is trying to update it but fails to do so

**Comment by R-N** [comment]
Wait up ill take a look in the code.

**Comment by monhustla** [comment]
Thank you buddy, it's very helpful.

**Comment by monhustla** [comment]
I just want to be clear, it never asks me for pin validation again when I rerun the code.

**Comment by R-N** [comment]
Yeah, that's weird. I think I needed to do pincode validation everytime I login with email & password. But well its been a long time since then, maybe I'm mistaken.

Change these lines :
if msg.type == 1:
            self.certificate = msg.certificate
            self.authToken = self._headers['X-Line-Access'] = msg.authToken

To :
 if msg.type == 1:
            If msg.certificate is not None and msg.certificate.strip():
                  self.certificate = msg.certificate
            self.authToken = self._headers['X-Line-Access'] = msg.authToken

Sorry for the bad indentation, I'm on phone

**Comment by R-N** [comment]
Only "self.certificate = msg.certificate" should be in the new if statement

**Comment by monhustla** [comment]
Does this look correct?

![img_0717](https://cloud.githubusercontent.com/assets/20822555/26765715/83dca490-4947-11e7-8c8b-99690a62e271.JPG)

**Comment by monhustla** [comment]
(sorry about the sideways images, my snipping tool doesn't display code in python for some reason)

**Comment by R-N** [comment]
Yes, that's correct, I think.

Try running it and make it blast messages like it did before

**Comment by monhustla** [comment]
I tried it and it returned the same error. It sent the message still but did not allow the code to loop.

**Comment by monhustla** [comment]
You mentioned the empty "certificate" string. The certificate variable in api.py is = "" . Is that something I need to change?

**Comment by monhustla** [comment]
This error is coming from line 68 of models.py also, not api.

**Comment by monhustla** [comment]
Sorry about the spam but here is the error in models.py 

![img_0718](https://cloud.githubusercontent.com/assets/20822555/26765820/7fd0963e-4949-11e7-8c57-62c653ace77c.JPG)

**Comment by R-N** [comment]
Sorry, I actually wasnt sure if logging in with authToken is missing the certificate or not.

Try also saving the certificate from the first client to a variable, then set it to the second client you made with authToken

client = LineClient(id, pw)
authToken = client.authToken
cert = client.certificate
prov = client.provider
mac = client.is_mac
client = LineClient(authToken=authToken)
client.id = id
client.password = pw
client.provider=prov
client.is_mac = mac
if client.certificate is None or not client.certificate.strip():
     client.certificate = cert

**Comment by monhustla** [comment]
Do those steps in API.py or my actual code?

**Comment by R-N** [comment]
Your actual code

**Comment by R-N** [comment]
Yeah it says the error came from models.py because it catches the exception then rethrows it

**Comment by R-N** [comment]
I dont think except-raise modifies the callstack traceback though

**Comment by monhustla** [comment]
Okay, this time it was different. It sent the messages but returned this error:
<img width="506" alt="object has no built in len" src="https://cloud.githubusercontent.com/assets/20822555/26765892/64858df6-494b-11e7-94b4-b2706e4f9cce.PNG">

@R-N

**Comment by monhustla** [comment]
Here is my code:
<img width="428" alt="results" src="https://cloud.githubusercontent.com/assets/20822555/26765904/bcbaeed0-494b-11e7-9496-93132b4c22c4.PNG">

**Comment by monhustla** [comment]
<img width="387" alt="group" src="https://cloud.githubusercontent.com/assets/20822555/26765909/e32c5ae0-494b-11e7-9e2f-ee127410389e.PNG">

**Comment by R-N** [comment]
Uh, please change the "id" to your email or whatever you signed in with.

Also, you havent censored your second password there.

Btw, I suggest that you remove the try-except in sendMessage which except block only re-raise the exception

try:
     return self....
except Exception as e:
      raise e

To just:
return self...

**Comment by monhustla** [comment]
Okay, I'll update that. Now I'm getting talk exception: 

![image](https://cloud.githubusercontent.com/assets/20822555/26765980/0429194e-494d-11e7-89aa-da70c0b262f8.png)

**Comment by monhustla** [comment]
Here is what I'm getting:, talkexception: 'not a member', code=10

![image](https://cloud.githubusercontent.com/assets/20822555/26765999/72812d46-494d-11e7-815f-650fbbc28dbe.png)

**Comment by monhustla** [comment]
It still sent the messages though

**Comment by R-N** [comment]
Ok, its email then.

Perhaps theres a pending group invitation? 
client.groups returns all of your groups, ones that you have joined and ones that is still just an invitation

**Comment by monhustla** [comment]
That may be it. I need to set the code to only send to groups I'm a member of. That would explain why it works well with friends but not groups. Only thing is I removed all of the groups that I'm a member of and don't see any invitations.

**Comment by monhustla** [comment]
It worked! Amazing help @R-N !

**Comment by R-N** [comment]
Wait, what worked lol
What did you do?

**Comment by monhustla** [comment]
I changed the code to accept group invites first, then send message. It looped just fine! Your changes also helped get past line logging me out. I really appreciate the help bud.

**Comment by R-N** [comment]
Nice.

Wonder why it doesnt show any pending group invitation in your Line (or did I misunderstood lol)

You're welcome :)

**Comment by monhustla** [comment]
I missed something small. Can I compensate you lol for your time?

**Comment by Nilpo** [comment]
You should really be storing your login credentials in an environmental variable and loading them from there so that they aren't hardcoded into your project. (Makes pasting your code a whole lot easier too!)

**Comment by anysz** [comment]
@Nilpo its great idea m8, but why environment var? why not just add somechange to api.py and client.py and load cert with email and pass like `cl = LineClient(id='EMAIL', password='PASS', cert='CERT')`

and for account kicked if we too many send data. i think @monhustla you can do it without scheduler by changing the server to 'https://gfs.line.naver.jp' and use /P4 or /S4 to Polling cause gd2 is line base server, it might had some patch. the original api.py is polling with /api/v4/TalkService.do which that url is just for getting authToken.

**Comment by Nilpo** [comment]
@anysz Most people are deploying to Heroku from GitHub. Using Environment Variables means you don't have to have your credentials stored in your project repo in plain text.

**Comment by anysz** [comment]
kk got it m8 @Nilpo
